V1 Topic Pages (rebased branch)

vanitabarrett opened the pull request:
Trello: https://trello.com/c/231ueER9/151-deploy-the-v1-topic-page

This branches off the original topic pages feature branch https://github.com/alphagov/collections/pull/538 and has been rebased against master.

<img width="1003" alt="screen shot 2018-03-23 at 11 31 22" src="https://user-images.githubusercontent.com/29889908/37827049-c015cc8a-2e8d-11e8-8e6e-6de157f7bd9d.png">

# Things to Check

- Topic pages still look as expected
- Compare between this PR and the original feature branch PR to ensure we haven't lost any commits
- World and Step by Step pages still look as expected
- Some browser testing of topic pages

# Links
Link to component guide: https://govuk-collections-pr-586.herokuapp.com/component-guide
Link to topic page root: 
- https://govuk-collections-pr-586.herokuapp.com/education
- https://govuk-collections-pr-586.herokuapp.com/childcare-parenting
- https://govuk-collections-pr-586.herokuapp.com/money

steventux commented on app/models/rummager_search.rb:
```diff
@@ -32,6 +41,20 @@ def start
+  def organisations
+    @_organisations ||= search_result.dig('aggregates', 'organisations', 'options').map do |option|
+      organisation = option['value']
+      Organisation.new(
```
Given the attribute names of the `Organisation` class are identical to the keys here you could probably abbreviate this to:
```ruby
attr_keys = %w(title content_id link slug organisation_state documents)
Organisation.new(organisation.slice(*attr_keys))
```
or similar.

fmlharrison replied:
The keys are slightly different than the attr name. I could change the attr name 'document_count' to 'documents'. 
But the organisation variable does not have the document key value pair on it so the above code does not quite work

steventux replied:
```ruby
keys = %w(title content_id link slug organisation_state)
attrs = organisation.slice(*keys).merge('document_count' => option['documents'])
Organisation.new(attrs)
```
or similar, though perhaps this is getting overly complex, you might be better off just being explicit about it.

fmlharrison replied:
I was going to do the above then I thought it was getting over complicated to read, so I am going to leave it as is.

steventux commented on test/models/organisation_test.rb:
```diff
@@ -0,0 +1,15 @@
+
+describe Organisation do
+  it 'returns the state of an organisation' do
+    organisation = Organisation.new(
```
Not all of these attrs are required for the test

```ruby
assert Organisation.new(organisation_state: 'live').live?, "Expected org to be live"
```

vanitabarrett replied:
Good point, now amended: https://github.com/alphagov/collections/pull/586/commits/e3a7756564a5bdabd0def6bb1669dab3329664ac#diff-207daaac6cfd16d598218e9fcc875105R6

steventux commented on test/services/most_popular_content_test.rb:
```diff
@@ -14,7 +14,7 @@ def taxon_content_id
 
   describe '#fetch' do
-    it 'returns the results from search, sorted by title' do
+    it 'returns the results from search' do
```
Given the commit message we're not testing what the amended ordering is here - not sure if this is covered elsewhere or in another commit.

steventux commented on test/presenters/taxon_presenter_test.rb:
```diff
@@ -216,6 +216,14 @@
 
   describe 'guidance_and_regulation_section' do
+    it 'checks whether guidance section should be shown' do
+      taxon = mock
```
I'd use a `before` block for this setup and a `let` block for `TaxonPresenter.new(taxon)`.

steventux commented on app/services/most_popular_content.rb:
```diff
@@ -1,33 +1,38 @@
 private
 
   def search_response
+    search_fields = %w(title
```
If this list isn't modified it could be a class constant.

steventux commented on app/services/most_recent_content.rb:
```diff
@@ -0,0 +1,37 @@
+private
+
+  def search_response
+    search_fields = %w(title
```
This is identical to the array of fields above in MostPopularContent (line 22) consider making a constant available to both.

steventux commented on test/services/tagged_organisations_test.rb:
```diff
@@ -0,0 +1,69 @@
+describe TaggedOrganisations do
+  describe '#fetch' do
+    it 'return tagged organisations' do
+      search_results = {
```
Seems like this could be refactored into either a helper method or `let` vars.

vanitabarrett replied:
Done https://github.com/alphagov/collections/pull/586/files#diff-62314a7230a7329a26b5cb721a328944

steventux commented on test/services/tagged_organisations_test.rb:
```diff
@@ -0,0 +1,69 @@
+    end
+
+    it 'only returns live organisations' do
+      search_results = {
```
Similarly, these tests span many lines due to the dependency setup, consider using `before` and `let` to break them up into setup and assertion.

vanitabarrett replied:
Done https://github.com/alphagov/collections/pull/586/files#diff-62314a7230a7329a26b5cb721a328944

steventux commented on test/support/rummager_helpers.rb:
```diff
@@ -18,15 +18,22 @@ def stub_content_for_taxon(content_ids, results, filter_navigation_document_supe
 
-  def stub_most_popular_content_for_taxon(content_id, results, filter_navigation_document_supertype: 'guidance')
+  def stub_most_popular_content_for_taxon(content_id, results, filter_content_purpose_supergroup: 'guidance_and_regulation')
+    fields = %w(title
```
This could be a seperate method,  it may even be able to service the same need in  `stub_most_recent_content_for_taxon`

steventux commented on app/presenters/taxon_presenter.rb:
```diff
@@ -1,81 +1,101 @@
-  def renders_as_accordion?
-    rendering_type == ACCORDION
+  def section_document_list(supergroup)
+    section_content(supergroup).each.map do |document|
```
I think we're conflating a few different concerns here, not necessary for this PR but I'd like to see a distinct class or module handling how supergroups generate their title and content. Perhaps a bunch of subclasses or something a bit more encapsulating of the supergroup specific logic. Seems to be spread across a few different classes here.

vanitabarrett commented on test/models/taxon_test.rb:
```diff
@@ -96,5 +79,44 @@
     end
+
+    it "requests news_and_communications content" do
+      results = %i(result_1 result_2 result_3 result_4 result_5)
```
Done https://github.com/alphagov/collections/blob/7a78fa79ac05867cbfa41ddb2cf801c4e28409dc/test/models/taxon_test.rb#L7

vanitabarrett commented:
@steventux When you get a chance, can you re-review? I think we addressed most of your comments, and the others will be written up in a story to revisit once this PR is merged in.
